roboteam_ai/src/world/Robot.cpp

```diff
@@ -89,21 +91,72 @@ double Robot::findBallDistance(const Vector2 &ballPos) {
     }
 
     // if the robot does not have ball
-    return -1.0;
+    return - 1.0;
+}
+
+const unsigned long Robot::getLastUpdatedWorldNumber() const {
+    return lastUpdatedWorldNumber;
 }
 
 unsigned char Robot::getGenevaState() const {
-    return genevaState[id];
+    return genevaState;
 }
 
 void Robot::setGenevaState(unsigned char state) {
-    if (id != -1 && state > 0 && state < 6) genevaState[id] = state;
-    else std::cout << "setting invalid geneva state (" << state << ") for robot with id " << id << std::endl;
+
+    if (state < 0 || state > 5) {
+        std::cout << "setting invalid geneva state (" << (int)state <<
+        ") for robot with id " << id << std::endl;
+    }
+    else if (! workingGeneva) {
+        std::cout << "setting geneva state (" << (int)state <<
+        ") for robot without working geneva with id " << id << std::endl;
+    }
+    else if (state != 0) {
+        previousGenevaState = genevaState;
+        genevaState = state;
+        timeGenevaChanged = world::world->getTime();
+    }
 }
 
-    const unsigned long &Robot::getLastUpdatedWorldNumber() const {
-        return lastUpdatedWorldNumber;
+bool Robot::isGenevaReady() const {
+    return world->getTime() - timeGenevaChanged >
+    abs(genevaState - previousGenevaState)*timeToChangeOneGenevaState;
+}
+
+bool Robot::hasWorkingGeneva() const {
+    return workingGeneva;
+}
+
+unsigned char Robot::getDribblerState() const {
+    return dribblerState;
+}
+
+bool Robot::isDribblerReady() const {
+    return world->getTime() - timeDribblerChanged >
```

Comment:
same as above.

Reply:
I don't like this function because this is not really how the dribbler works. We don't have to wait for a dribbler to change it's level.

Reply:
it is a check to see if the ball had time to match the speed of the dribbler